### PR TITLE
mgr/volumes: prevent negative subvolume size

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: 
```
$ ./bin/ceph fs subvolume create myfs mysubvol --size -10 --group_name mygroup --pool_layout mycephfs_data --mode 777
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2019-09-01T21:42:23.406+0530 7f08ce9f6700 -1 WARNING: all dangerous and experimental features are enabled.
2019-09-01T21:42:23.503+0530 7f08ce9f6700 -1 WARNING: all dangerous and experimental features are enabled.
Error EINVAL: Traceback (most recent call last):
  File "/home/jcollin/workspace/ceph/src/pybind/mgr/mgr_module.py", line 915, in _handle_command
    return self.handle_command(inbuf, cmd)
  File "/home/jcollin/workspace/ceph/src/pybind/mgr/volumes/module.py", line 188, in handle_command
    return handler(inbuf, cmd)
  File "/home/jcollin/workspace/ceph/src/pybind/mgr/volumes/module.py", line 230, in _cmd_fs_subvolume_create
    mode=cmd.get('mode', '755'))
  File "/home/jcollin/workspace/ceph/src/pybind/mgr/volumes/fs/volume.py", line 345, in conn_wrapper
    result = func(self, fs_h, **kwargs)
  File "/home/jcollin/workspace/ceph/src/pybind/mgr/volumes/fs/volume.py", line 372, in create_subvolume
    sv.create_subvolume(spec, size, pool=pool, mode=self.octal_str_to_decimal_int(mode))
  File "/home/jcollin/workspace/ceph/src/pybind/mgr/volumes/fs/subvolume.py", line 111, in create_subvolume
    raise e
InvalidValue: error in setxattr: Invalid argument [Errno 22]
```

### CODE REVIEW

**Comment by ajarr** [comment]
Please add a test

**Comment by ajarr** [comment]
As this is a standalone PR, can you add a tracker ticket so that we don't miss backporting this to nautilus?

**Comment by joscollin** [comment]
@ajarr `create_subvolume` can be called with `0` size now. Let me know if you want to avoid it.

**Comment by ajarr** [comment]
> @ajarr `create_subvolume` can be called with `0` size now. Let me know if you want to avoid it.

Does setxattr let you do that?

File: src/pybind/mgr/volumes/fs/subvolume.py [hunk 1]
```diff
@@ -77,6 +77,9 @@ def create_subvolume(self, spec, size=None, namespace_isolated=True, mode=0o755,
 
         try:
             if size is not None:
+                if size < 0:
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/subvolume.py]: Instead can we catch the InvalidValue exception raised by setxattr, and raise a VolumeException with an appropriate message?

Comment (joscollin) [reply src/pybind/mgr/volumes/fs/subvolume.py]: > Does setxattr let you do that?

Yes `setxattr` was allowing `0`. But it is fixed now.

Comment (joscollin) [reply src/pybind/mgr/volumes/fs/subvolume.py]: @ajarr Pushed the fix. The testing shows that this code still allows setting size `0`, but prevents negative values.

File: qa/tasks/cephfs/test_volumes.py [hunk 2]
```diff
@@ -162,6 +162,15 @@ def test_subvolume_create_with_auto_cleanup_on_fail(self):
         else:
             raise
 
+    def test_subvolume_create_with_invalid_size(self):
+        # create subvolume with an invalid size -1
+        subvolume = self._generate_random_subvolume_name()
+        try:
+            self._fs_cmd("subvolume", "create", self.volname, subvolume, "--size", -1)
+        except CommandFailedError as ce:
+            if ce.exitstatus != errno.ENOENT:
```
Comment (ajarr) [inline qa/tasks/cephfs/test_volumes.py]: s/errno.ENOENT/errno.EINVAL/ ?

File: qa/tasks/cephfs/test_volumes.py [hunk 3]
```diff
@@ -162,6 +162,15 @@ def test_subvolume_create_with_auto_cleanup_on_fail(self):
...
+                raise
```
Comment (ajarr) [inline qa/tasks/cephfs/test_volumes.py]: Also add,
```
else:
        raise
```

The test should error out if no exception is raised

File: qa/tasks/cephfs/test_volumes.py (hunk 2 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_volumes.py]: Then it won't raise here I think, as we are raising EINVAL in  [subvolume.py](https://github.com/ceph/ceph/pull/30058/files#diff-9455d141cc54cc29b319d9e6a29678c8R83)

File: qa/tasks/cephfs/test_volumes.py (hunk 3 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_volumes.py]: Yes, we can raise an exception here. But using raise without arguments shows an error "TypeError: exceptions must be old-style classes or derived from BaseException, not NoneType". Please see https://stackoverflow.com/a/27947646

Comment (ajarr) [reply qa/tasks/cephfs/test_volumes.py]: Good point. Raise RuntimeError with an error message that you expected fs subvolume create to fail here?
I think there are other places in this code where we incorrectly use `raise` when not re-raising Exceptions. We should clean that up in this or a later PR. Maybe just file a tracker ticket for now?
```
else:
    raise RuntimeError("expected the `fs subvolume create` command to fail")
 ```

File: qa/tasks/cephfs/test_volumes.py (hunk 2 above)
Comment (ajarr) [reply qa/tasks/cephfs/test_volumes.py]: The exit status should be EINVAL, right? ENOENT doesn't make sense here. Maybe you've to change the code in subvolume.py,
```
raise VolumeException(-e.args[0], "Invalid size: '{0}'".format(size))
```
 to
```
raise VolumeException(-errno.EINVAL, "Invalid size: '{0}'".format(size))
```

File: qa/tasks/cephfs/test_volumes.py (hunk 3 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_volumes.py]: I already have a PR that was closed. I will re-open it and will do the needful.

**Comment by ajarr** [comment]
@batrick please take a look

**Comment by joscollin** [comment]
Test Results:

[1] https://shaman.ceph.com/builds/ceph/wip-negative-subvol-size/
[2] http://pulpito.front.sepia.ceph.com/jcollin-2019-09-13_15:07:46-fs-wip-negative-subvol-size-distro-basic-smithi/


```
2019-09-13T15:51:23.272 INFO:tasks.cephfs_test_runner:Starting test: test_subvolume_create_with_invalid_size (tasks.cephfs.test_volumes.TestVolumes)
2019-09-13T15:51:23.272 INFO:teuthology.orchestra.run.smithi053:Running:
2019-09-13T15:51:23.272 INFO:teuthology.orchestra.run.smithi053:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph log 'Starting test\
 tasks.cephfs.test_volumes.TestVolumes.test_subvolume_create_with_invalid_size'
...


2019-09-13T15:51:47.170 INFO:teuthology.orchestra.run.smithi053.stderr:2019-09-13T15:51:47.164+0000 7f3de12c8700  1 -- 172.21.15.53:0/1774105653 --> [v2:172.21.15.134:6800/12039,v1:172.21.1\
5.134:6801/12039] -- mgr_command(tid 0: {"prefix": "fs subvolume create", "vol_name": "cephfs", "sub_name": "subvolume_8172", "target": ["mgr", ""], "size": -1}) v1 -- 0x7f3ddc05a230 con 0x\
7f3dc00252b0
...

2019-09-13T15:51:47.201 INFO:tasks.ceph.mgr.x.smithi134.stderr:2019-09-13T15:51:47.198+0000 7f826e3a6700 -1 mgr.server reply reply (22) Invalid argument Invalid size: '-1'
2019-09-13T15:51:47.202 INFO:teuthology.orchestra.run.smithi053.stderr:2019-09-13T15:51:47.196+0000 7f3dd3fff700  1 -- 172.21.15.53:0/1774105653 <== mgr.4108 v2:172.21.15.134:6800/12039 1 =\
=== mgr_command_reply(tid 0: -22 Invalid size: '-1') v1 ==== 26+0+0 (crc 0 0 0) 0x7f3ddc05a230 con 0x7f3dc00252b0
2019-09-13T15:51:47.234 INFO:teuthology.orchestra.run.smithi053.stderr:Error EINVAL: Invalid size: '-1'
...

2019-09-13T15:51:47.250 DEBUG:teuthology.orchestra.run:got remote process result: 22
2019-09-13T15:51:47.251 INFO:teuthology.orchestra.run.smithi053:Running:
2019-09-13T15:51:47.251 INFO:teuthology.orchestra.run.smithi053:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph log 'Ended test ta\
sks.cephfs.test_volumes.TestVolumes.test_subvolume_create_with_invalid_size'
...

2019-09-13T15:52:18.368 INFO:tasks.cephfs_test_runner:test_subvolume_create_with_invalid_size (tasks.cephfs.test_volumes.TestVolumes) ... ok


```

**Pull request merged**